effect/jvm/src/test/scala/scalaz/effect/generator/GenIO.scala

```diff
@@ -0,0 +1,150 @@
+   */
+  def genAsyncSuccess[E, A: Arbitrary]: Gen[IO[E, A]] =
+    for {
+      a <- Arbitrary.arbitrary[A]
+    } yield IO.async[E, A](cb => cb(ExitResult.Completed(a)))
+
+  /**
+   * Randomly uses either `genSyncSuccess` or `genAsyncSuccess` with equal probability.
+   */
+  def genSuccess[E, A: Arbitrary]: Gen[IO[E, A]] = Gen.oneOf(genSyncSuccess[E, A], genAsyncSuccess[E, A])
+
+  /**
+   * Given a generator for `E`, produces a generator for `IO[E, A]` using the `IO.fail` constructor.
+   */
+  def genSyncFailure[E: Arbitrary, A]: Gen[IO[E, A]] =
+    for {
+      error <- Arbitrary.arbitrary[E]
+    } yield IO.fail(error)
+
+  /**
+   * Given a generator for `E`, produces a generator for `IO[E, A]` using the `IO.async` constructor.
+   */
+  def genAsyncFailure[E: Arbitrary, A]: Gen[IO[E, A]] =
+    for {
+      error <- Arbitrary.arbitrary[E]
+    } yield IO.async(cb => cb(ExitResult.Failed(error)))
+
+  /**
+   * Randomly uses either `genSyncFailure` or `genAsyncFailure` with equal probability.
+   */
+  def genFailure[E: Arbitrary, A]: Gen[IO[E, A]] = Gen.oneOf(genSyncFailure[E, A], genAsyncFailure[E, A])
+
+  /**
+   * Randomly uses either `genSuccess` or `genFailure` with equal probability.
+   */
+  def genIO[E: Arbitrary: Cogen, A: Arbitrary: Cogen]: Gen[IO[E, A]] =
+    Gen.oneOf(genSuccess[E, A], genFailure[E, A])
+
+  /**
+   * Given a generator for `IO[E, A]`, produces a sized generator for `IO[E, A]` which represents a transformation,
+   * by using some random combination of the methods `map`, `flatMap`, `leftMap`, and any other method that does not change
+   * the success/failure of the value, but may change the value itself.
+   */
+  def genLikeTrans[E: Arbitrary: Cogen, A: Arbitrary: Cogen](gen: Arbitrary[IO[E, A]]): Gen[IO[E, A]] = {
+    val successFlatMapGen = genOfFlatMaps[E, A](Cogen[A], Arbitrary(genSuccess))
+    genTransformations(Gen.oneOf(successFlatMapGen, genOfMaps[A], genOfLeftMaps[E]))(gen)
+  }
+
+  /**
+   * Given a generator for `IO[E, A]`, produces a sized generator for `IO[E, A]` which represents a transformation,
+   * by using methods that can have no effect on the resulting value (e.g. `map(identity)`, `io.race(never)`, `io.par(io2).map(_._1)`).
+   */
+  def genIdentityTrans[E: Arbitrary: Cogen, A: Arbitrary: Cogen](gen: Arbitrary[IO[E, A]]): Gen[IO[E, A]] =
+    genTransformations(
+      Gen.oneOf(genOfIdentityFlatMaps[E, A],
+                genOfIdentityMaps[A],
+                genOfIdentityLeftMaps[E],
+                genOfRace[E, A],
+                genOfParallel[E, A](Arbitrary(genSuccess)))
+    )(gen)
+
+  private def genTransformations[E: Arbitrary: Cogen, A: Arbitrary: Cogen](
+    functionsGen: Gen[FunctionCall]
+  )(gen: Arbitrary[IO[E, A]]): Gen[IO[E, A]] =
+    Gen.sized { size =>
+      for {
+        io        <- gen.arbitrary
+        functions <- Gen.listOfN(size, functionsGen)
```

Comment:
If you want to simplify this, you actually do not need the intermediate structure `FunctionCall`.

You can instead have a recursive function which stops appending operations when it reaches `n <= 0`.

Something like:

```scala
def append1(n: Int, io: IO[E, A]): Gen[IO[E, A]] = 
  if (n <= 0) io
  else (for {
    io <- Gen.oneOf(genFlatMap(io), genLeftMap(io), ...)
  } yield io).flatMap(append1(n - 1, _))
```

Then call this with `sized` as the starting condition.